src/main/ruby/core/process_mirror.rb

```diff
@@ -264,26 +287,32 @@ def convert_to_fd(obj, target)
           when ::IO
             obj.fileno
           when ::String
-            [obj, default_mode(target), 0644]
+            open_file_for_child(file_list, obj, default_mode(target), 0644)
           when ::Array
             case obj.size
             when 1
-              [obj[0], File::RDONLY, 0644]
+              open_file_for_child(file_list, obj[0], File::RDONLY, 0644)
             when 2
               if obj[0] == :child
-                fd = convert_to_fd obj[1], target
+                fd = convert_to_fd obj[1], target, file_list
                 fd.kind_of?(::Fixnum) ?  -(fd + 1) : fd
               else
-                [obj[0], convert_file_mode(obj[1]), 0644]
+                open_file_for_child(file_list, obj[0], convert_file_mode(obj[1]), 0644)
               end
             when 3
-              [obj[0], convert_file_mode(obj[1]), obj[2]]
+              open_file_for_child(file_list, obj[0], convert_file_mode(obj[1]), obj[2])
             end
           else
             raise ArgumentError, "wrong exec redirect: #{obj.inspect}"
           end
         end
 
+        def open_file_for_child(file_list, name, mode, perms)
+          f = File.new(name, mode, perms)
+          file_list << f
```

Comment:
Is it needed to pass the `file_list` around? Otherwise I think using `@files_for_child` directly would be clearer, it's part of the state of the spawn setup.